clouddrift/wavelet.py

```diff
@@ -0,0 +1,518 @@
+        If the shape of time axis is different for input and wavelet.
+        If ``boundary not in ["mirror", "zeros", "periodic"]``.
+
+    See Also
+    --------
+    :func:`morsewave`, `morsefreq`
+    """
+    # time_axis must be in valid range
+    if time_axis < -1 or time_axis > len(x.shape) - 1:
+        raise ValueError(
+            f"time_axis ({time_axis}) is outside of the valid range ([-1,"
+            f" {len(x.shape) - 1}])."
+        )
+    # Positions and time arrays must have the same shape.
+    if x.shape[time_axis] != wave.shape[-1]:
+        raise ValueError("x and wave time axes must have the same length.")
+
+    wave_ = np.moveaxis(wave, [freq_axis, order_axis], [-2, -3])
+
+    # if x is of dimension 1 we need to expand
+    # otherwise make sure time axis is last
+    if np.ndim(x) < 2:
+        x_ = np.expand_dims(x, axis=0)
+    else:
+        x_ = np.moveaxis(x, time_axis, -1)
+
+    if ~np.all(np.isreal(x)):
+        if norm == "energy":
+            x_ /= np.sqrt(2)
+        elif norm == "bandpass":
+            x_ /= 2
+
+    # to do: add detrending option by default
+
+    # apply boundary conditions
+    if boundary == "mirror":
+        x_ = np.concatenate((np.flip(x_, axis=-1), x_, np.flip(x_, axis=-1)), axis=-1)
+    elif boundary == "zeros":
+        x_ = np.concatenate((np.zeros_like(x_), x_, np.zeros_like(x_)), axis=-1)
+    elif boundary == "periodic":  # JML: this not needed in this case you just want n
+        x_ = np.concatenate((x_, x_, x_), axis=-1)
+    else:
+        raise ValueError("boundary must be one of 'mirror', 'align', or 'zeros'.")
+
+    time_length = np.shape(x)[-1]
+    time_length_ = np.shape(x_)[-1]
+
+    # pad wavelet with zeros: JML ok
+    order_length, freq_length, _ = np.shape(wave)
+    _wave = np.zeros((order_length, freq_length, time_length_), dtype=np.cdouble)
+
+    index = slice(
+        int(np.floor(time_length_ - time_length) / 2),
+        int(time_length + np.floor(time_length_ - time_length) / 2),
+    )
+    _wave[:, :, index] = wave_
+
+    # take fft along axis = -1
+    _wavefft = np.fft.fft(_wave)
+    om = 2 * np.pi * np.linspace(0, 1 - 1 / time_length_, time_length_)
+    if time_length_ % 2 == 0:
+        _wavefft = (
+            _wavefft * np.exp(1j * -om * (time_length_ + 1) / 2) * np.sign(np.pi - om)
+        )
+    else:
+        _wavefft = _wavefft * np.exp(1j * -om * (time_length_ + 1) / 2)
+
+    # here we should be able to automate the tiling without assuming extra dimensions of wave
+    X_ = np.tile(
+        np.expand_dims(np.fft.fft(x_), (-3, -2)),
+        (1, order_length, freq_length, 1),
+    )
+
+    # finally the transform; return precision of input `x``; central part only
+    complex_dtype = np.cdouble if x.dtype == np.single else np.csingle
+    wtx = np.fft.ifft(X_ * np.conj(_wavefft)).astype(complex_dtype)
+    wtx = wtx[..., index]
+    # remove extra dimensions
+    wtx = np.squeeze(wtx)
+
+    return wtx
+
+
+def morsewave(
```

Comment:
What do you think about

```suggestion
def morse_wavelets(
```